src/reduce/mapreduce_1d.jl

```diff
@@ -125,14 +125,14 @@ function mapreduce_1d(
     blocks = (len + num_per_block - 1) ÷ num_per_block
 
     if !isnothing(temp)
-        @argcheck get_backend(temp) === get_backend(src)
+        @argcheck get_backend(temp) === backend
         @argcheck eltype(temp) === typeof(init)
         @argcheck length(temp) >= blocks * 2
         dst = temp
     else
         # Figure out type for destination
         dst_type = typeof(init)
-        dst = similar(src, dst_type, blocks * 2)
+        dst = KernelAbstractions.allocate(backend, dst_type, blocks * 2)
```

Comment:
Any reason in particular for this change? This will make the output storage mode for GPU arrays always be the default instead of matching the input array. (Same goes for all the other similar changes in the other file)

Reply:
It is to allow using unmaterialised ranges as inputs (see [issue](https://github.com/JuliaGPU/AcceleratedKernels.jl/issues/33)) - what other GPU arrays could we have besides the defaults?

Reply:
Just saw https://github.com/JuliaGPU/AcceleratedKernels.jl/issues/33. I think wrong storage mode is a less disruptive bug than what it fixes so maybe this should stay for now.

Metal has Private (default in Metal.jl), Shared, and Managed storage, with private being only accessible to the GPU, and Shared being accessible to GPU and CPU.

CUDA has DeviceMemory, UnifiedMemory, and HostMemory, which I assume behave similarly to the Metal ones.

Reply:
Hmm, maybe we should have some sort of UnmaterialisedBackend for cases like this - then we could distinguish between ranges and specialised backend arrays (@vchuravy ?); for the moment, `temp` could be specified explicitly for a given storage

Reply:
Still, if the default storage mode is GPU-only, wouldn't that be the best storage for an unspecified temporary array? We can move data between a Private storage to a Shared one, right?